Implement semaphores

dkochmanski opened the pull request:
Includes generic implementation which depends on condition variables and locks.

CCL, ECL and SBCL have native implementation (ECL implementation will trigger on next release). Default implementation was tested on all three as well. Test suite is added.

dkochmanski commented:
@scymtym, @xrme: could you please help review this PR? Implementation for both SBCL and CCL is minimal, generic implementation is also relatively short.

@sionescu: test suite checking functionality is included in PR and all tests pass AFAICT.

xrme commented:
The use of CCL semaphore functions looks OK to me.

dkochmanski commented:
> create-semaphore: I don't like the name and I don't understand the argument. Why is being able to :use :bt in cl-user a requirement?

You are right, I'm not sure what I was thinking (I remember debugger popped up when loaded bordeaux-threads on ccl, but that had to be something else, because I can't reproduce it. Changing name in a minute.

> create-semaphore: Is it possible to add an optional (or &key, rather) parameter for specifying the initial count? if not, please explain the reason in the commit message.

I did that at first, but to match `wait-on-semaphore` I've decided not to. I'm fine with both options.

> Could wait-on-semaphore accept an optional count argument? This would be symmetric w.r.t. signal-semaphore and seems to be common practice in semaphore APIs.

- `wait-on-semaphore` with optional count argument can't be used portably (neither CCL nor ECL supports it, so we'd have to signal them in a loop to achieve portability, and that may lead to starvation of waiters, who require more than one from the counter for crowded systems) – since BT is the least common denominator it is only fair to accept this limitation

dkochmanski commented:
Changes pushed. Also adjusted tests to take into account `:count` argument to `make-semaphore`.

scymtym commented:
Maybe remove the comment regarding the name ``create-semaphore`` from the commit message?

dkochmanski commented:
done, thanks.

sionescu commented on src/default-implementations.lisp:
```diff
@@ -268,6 +268,51 @@ WITH-LOCK-HELD etc etc"
+
+(defdfun make-semaphore (&key name (count 0))
+  "Create a semaphore with the supplied NAME with initial counter value COUNT."
+  (list :lock (make-lock name)
```
Please turn this into a struct, and add a type predicate as with the rest of the API.

dkochmanski replied:
How do you want me to do that? We generally don't define structures in `default-implementations.lisp`. We do something like that for lispworks in `impl-lispworks-condition-variables`, and for digitool in `condition-variables`, but it is implementation-specific.

In `default-implementation` we cons a list in `make-lock` and return NIL for other structures.

We do `define-condition` for timeout on platforms other than SBCL in `bordeaux-threads` file. Should I do a similar thing in this file conditionalized with implemented applications so far?

```diff
diff --git a/src/bordeaux-threads.lisp b/src/bordeaux-threads.lisp
index 9749944..56fff95 100644
--- a/src/bordeaux-threads.lisp
+++ b/src/bordeaux-threads.lisp
@@ -43,6 +43,11 @@ Distributed under the MIT license (see LICENSE file)
                          (timeout-length c))
                  (format s "A timeout occurred.")))))
 
+#-(or ccl ecl sbcl)
+(defstruct %semaphore
+  lock
+  cv
+  counter)
 
 ;;; Thread Creation
 
```

below define-condition? above it? somewhere else?

sionescu commented on src/impl-ecl.lisp:
```diff
@@ -106,6 +106,26 @@ Distributed under the MIT license (see LICENSE file)
+
+       (defun wait-on-semaphore (semaphore &key timeout)
+         (if timeout
+             (handler-case (with-timeout (timeout)
```
Is the use of WITH-TIMEOUT safe ?

dkochmanski replied:
It won't crash nor anything. ECL has some limitation for timeouts – you can't interrupt sleeping process. So 

In sense of crashing it is safe (there should be nothing wrong with it). It has a drawback, because nesting timeouts doesn't work (you can't interrupt process which is during the call to SLEEP like that). Effectively if you do:

```
(with-timeout (1) (wait-on-semaphore … :timeout 2))
```

it will end after two seconds. Compare to much simpler:

```
CL-USER> (time (handler-case (bt:with-timeout (0.1) (sleep 1) "hi")
                   (bt:timeout () 3)))
real time : 1.001 secs
run time  : 0.000 secs
gc count  : 1 times
consed    : 5395424 bytes
```

That said, I want to add different kind of timeout (interruptible) to both semaphores and condition-variables, but I don't know if I will manage to do it before ECL 16.2.0 release.

dkochmanski commented:
I've decided, that I will postpone native ECL semaphore implementation until I have implemented missing timeouts for both condition variables and semaphores.

Changes: - add structure %semaphore for generic implementation (always defined to avoid compilation warnings when default implementations are compiled)
- added deftype for this structure
- changed keyword argument
- moved with-timeout to file bordeaux-threads
- fixed condition-variable-wait function for ECL (when timeout is provided)
- new test to verify, if semaphore is of type bt:semaphore (newly exported symbol)

dkochmanski commented:
Problem with travis comes from:
```
Installing alisp-10.0
Making symlink.
Adding alisp-10.0 configuration.
Installing quicklisp.
Failed to install quicklisp.
Lisp has expired. See http://www.franz.com/support/newlicense.
The command "curl -L https://raw.githubusercontent.com/sionescu/cl-travis/master/install.sh | sh" failed and exited with 1 during .
Your build has been stopped.
```

is there something more I should fix in this PR?

dkochmanski commented:
I have added type predicate. Is there something else what is missing or not addressed?

dkochmanski commented:
@sionescu I've added changes requested by the last pass of the review.

sionescu commented:
Merged, thank you.
